File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/ThirdPartyTransferFragment.java
package org.mifos.mobilebanking.ui.fragments;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatButton;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.beneficiary.Beneficiary;
import org.mifos.mobilebanking.models.beneficiary.BeneficiaryDetail;
import org.mifos.mobilebanking.models.payload.AccountDetail;
import org.mifos.mobilebanking.models.payload.TransferPayload;
import org.mifos.mobilebanking.models.templates.account.AccountOption;
import org.mifos.mobilebanking.models.templates.account.AccountOptionsTemplate;
import org.mifos.mobilebanking.presenters.ThirdPartyTransferPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.adapters.AccountsSpinnerAdapter;
import org.mifos.mobilebanking.ui.adapters.BeneficiarySpinnerAdapter;
import org.mifos.mobilebanking.ui.enums.TransferType;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.ThirdPartyTransferView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.DateHelper;
import org.mifos.mobilebanking.utils.MFDatePicker;
import org.mifos.mobilebanking.utils.Network;
import org.mifos.mobilebanking.utils.ProcessView;
import org.mifos.mobilebanking.utils.Toaster;
import org.mifos.mobilebanking.utils.Utils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by dilpreet on 21/6/17.
 */

public class ThirdPartyTransferFragment extends BaseFragment implements ThirdPartyTransferView,
        AdapterView.OnItemSelectedListener {

    @BindView(R.id.sp_beneficiary)
    Spinner spBeneficiary;

    @BindView(R.id.sp_pay_from)
    Spinner spPayFrom;

    @BindView(R.id.et_amount)
    EditText etAmount;

    @BindView(R.id.et_remark)
    EditText etRemark;

    @BindView(R.id.ll_make_transfer)
    LinearLayout layoutMakeTransfer;

    @BindView(R.id.process_one)
    ProcessView pvOne;

    @BindView(R.id.process_two)
    ProcessView pvTwo;

    @BindView(R.id.process_three)
    ProcessView pvThree;

    @BindView(R.id.process_four)
    ProcessView pvFour;

    @BindView(R.id.btn_pay_to)
    AppCompatButton btnPayTo;

    @BindView(R.id.btn_pay_from)
    AppCompatButton btnPayFrom;

    @BindView(R.id.btn_amount)
    AppCompatButton btnAmount;

    @BindView(R.id.btn_add_beneficiary)
    AppCompatButton btnAddBeneficiary;

    @BindView(R.id.ll_review)
    LinearLayout llReview;

    @BindView(R.id.tv_select_beneficary)
    TextView tvSelectBeneficiary;

    @BindView(R.id.tv_select_amount)
    TextView tvEnterAmount;

    @BindView(R.id.tv_enter_remark)
    TextView tvEnterRemark;

    @BindView(R.id.tv_add_beneficiary_msg)
    TextView tvAddBeneficiaryMsg;

    @BindView(R.id.layout_error)
    View layoutError;

    @Inject
    ThirdPartyTransferPresenter presenter;

    private List<BeneficiaryDetail> listBeneficiary = new ArrayList<>();
    private List<AccountDetail> listPayFrom = new ArrayList<>();
    private List<Beneficiary> beneficiaries;
    private BeneficiarySpinnerAdapter beneficiaryAdapter;
    private AccountsSpinnerAdapter payFromAdapter;
    private AccountOption fromAccountOption;
    private AccountOption beneficiaryAccountOption;
    private AccountOptionsTemplate accountOptionsTemplate;
    private String transferDate;
    private View rootView;
    private SweetUIErrorHandler sweetUIErrorHandler;

    public static ThirdPartyTransferFragment newInstance() {
        ThirdPartyTransferFragment fragment = new ThirdPartyTransferFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        rootView = inflater.inflate(R.layout.fragment_third_party_transfer, container, false);
        setToolbarTitle(getString(R.string.third_party_transfer));
        ButterKnife.bind(this, rootView);
        sweetUIErrorHandler = new SweetUIErrorHandler(getActivity(), rootView);
        showUserInterface();

        presenter.attachView(this);
        if (savedInstanceState == null) {
            presenter.loadTransferTemplate();
        }

        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(Constants.TEMPLATE, accountOptionsTemplate);
        outState.putParcelableArrayList(Constants.BENEFICIARY, new ArrayList<Parcelable>(
                beneficiaries));
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (savedInstanceState != null) {
            showThirdPartyTransferTemplate((AccountOptionsTemplate) savedInstanceState.
                    getParcelable(Constants.TEMPLATE));
            List<Beneficiary> tempBeneficiaries = savedInstanceState.getParcelableArrayList(
                    Constants.BENEFICIARY);
            showBeneficiaryList(tempBeneficiaries);
        }
    }

    /**
     * Setting up basic components
     */
    @Override
    public void showUserInterface() {
        payFromAdapter = new AccountsSpinnerAdapter(getActivity(), R.layout.account_spinner_layout,
                listPayFrom);
        payFromAdapter.setDropDownViewResource(android.R.layout.select_dialog_singlechoice);
        spPayFrom.setAdapter(payFromAdapter);
        spPayFrom.setOnItemSelectedListener(this);

        beneficiaryAdapter = new BeneficiarySpinnerAdapter(getActivity(),
                R.layout.beneficiary_spinner_layout, listBeneficiary);
        beneficiaryAdapter.setDropDownViewResource(android.R.layout.select_dialog_singlechoice);
        spBeneficiary.setAdapter(beneficiaryAdapter);
        spBeneficiary.setOnItemSelectedListener(this);

        transferDate = DateHelper.getSpecificFormat(DateHelper.FORMAT_dd_MMMM_yyyy,
                MFDatePicker.getDatePickedAsString());

        pvOne.setCurrentActive();
    }

    /**
     * Checks validation of {@code etRemark} and then opens {@link TransferProcessFragment} for
     * initiating the transfer
     */
    @OnClick(R.id.btn_review_transfer)
    public void reviewTransfer() {
        if (etAmount.getText().toString().equals("")) {
            Toaster.show(rootView, getString(R.string.enter_amount));
            return;
        }

        if (etAmount.getText().toString().equals(".")) {
            Toaster.show(rootView, getString(R.string.invalid_amount));
            return;
        }

        if (etRemark.getText().toString().trim().equals("")) {
            Toaster.show(rootView, getString(R.string.remark_is_mandatory));
            return;
        }

        if (spBeneficiary.getSelectedItem().toString().
                equals(spPayFrom.getSelectedItem().toString())) {
            Toaster.show(rootView, getString(R.string.error_same_account_transfer));
            return;
        }

        TransferPayload transferPayload = new TransferPayload();
        transferPayload.setFromAccountId(fromAccountOption.getAccountId());
        transferPayload.setFromClientId(fromAccountOption.getClientId());
        transferPayload.setFromAccountType(fromAccountOption.getAccountType().getId());
        transferPayload.setFromOfficeId(fromAccountOption.getOfficeId());
        transferPayload.setToOfficeId(beneficiaryAccountOption.getOfficeId());
        transferPayload.setToAccountId(beneficiaryAccountOption.getAccountId());
        transferPayload.setToClientId(beneficiaryAccountOption.getClientId());
        transferPayload.setToAccountType(beneficiaryAccountOption.getAccountType().getId());
        transferPayload.setTransferDate(transferDate);
        transferPayload.setTransferAmount(Double.parseDouble(etAmount.getText().toString()));
        transferPayload.setTransferDescription(etRemark.getText().toString());

        ((BaseActivity) getActivity()).replaceFragment(TransferProcessFragment.
                newInstance(transferPayload, TransferType.TPT), true, R.id.container);

    }

    /**
     * Shows a {@link android.support.design.widget.Snackbar} with {@code message}
     * @param msg String to be shown
     */
    @Override
    public void showToaster(String msg) {
        Toaster.show(rootView, msg);
    }

    /**
     * Provides with {@code accountOptionsTemplate} fetched from server which is used to update
     * {@code listPayFrom}
     * @param accountOptionsTemplate Template for account transfer
     */
    @Override
    public void showThirdPartyTransferTemplate(AccountOptionsTemplate accountOptionsTemplate) {
        this.accountOptionsTemplate = accountOptionsTemplate;
        listPayFrom.clear();
        listPayFrom.addAll(presenter.getAccountNumbersFromAccountOptions(accountOptionsTemplate.
                getFromAccountOptions()));
        payFromAdapter.notifyDataSetChanged();

    }

    /**
     * Provides with {@code beneficiaries} fetched from server which is used to update
     * {@code listBeneficiary}
     * @param beneficiaries List of {@link Beneficiary} linked with user's account
     */
    @Override
    public void showBeneficiaryList(List<Beneficiary> beneficiaries) {
        this.beneficiaries = beneficiaries;
        listBeneficiary.clear();
        listBeneficiary.addAll(presenter.getAccountNumbersFromBeneficiaries(beneficiaries));
        beneficiaryAdapter.notifyDataSetChanged();
    }


    /**
     *  Disables {@code spPayFrom} {@link Spinner} and sets {@code pvOne} to completed and make
     * {@code pvThree} pvTwo
     */
    @OnClick(R.id.btn_pay_from)
    public void payFromSelected() {
        pvOne.setCurrentCompeleted();
        pvTwo.setCurrentActive();

        btnPayFrom.setVisibility(View.GONE);

        tvSelectBeneficiary.setVisibility(View.GONE);
        if (!listBeneficiary.isEmpty()) {
            btnPayTo.setVisibility(View.VISIBLE);
            spBeneficiary.setVisibility(View.VISIBLE);
            spPayFrom.setEnabled(false);
        } else {
            tvAddBeneficiaryMsg.setVisibility(View.VISIBLE);
            btnAddBeneficiary.setVisibility(View.VISIBLE);
        }
    }


    /**
     * Checks validation of {@code spBeneficiary} {@link Spinner}.<br>
     *  Disables {@code spBeneficiary} {@link Spinner} and sets {@code pvTwo} to completed and make
     * {@code pvThree} active
     */
    @OnClick(R.id.btn_pay_to)
    public void payToSelected() {
        if (spBeneficiary.getSelectedItem().toString().equals(spPayFrom.getSelectedItem().
                toString())) {
            showToaster(getString(R.string.error_same_account_transfer));
            return;
        }
        pvTwo.setCurrentCompeleted();
        pvThree.setCurrentActive();

        btnPayTo.setVisibility(View.GONE);
        tvEnterAmount.setVisibility(View.GONE);
        etAmount.setVisibility(View.VISIBLE);
        btnAmount.setVisibility(View.VISIBLE);
        spBeneficiary.setEnabled(false);
    }

    /**
     * Checks validation of {@code etAmount} {@link EditText}.<br>
     * Disables {@code etAmount} and sets {@code pvThree} to completed and make
     * {@code pvFour} active
     */
    @OnClick(R.id.btn_amount)
    public void amountSet() {

        if (etAmount.getText().toString().equals("")) {
            showToaster(getString(R.string.enter_amount));
            return;
        }

        if (etAmount.getText().toString().equals(".")) {
            showToaster(getString(R.string.invalid_amount));
            return;
        }

        if (Double.parseDouble(etAmount.getText().toString()) == 0) {
            showToaster(getString(R.string.amount_greater_than_zero));
            return;
        }

        pvThree.setCurrentCompeleted();
        pvFour.setCurrentActive();

        btnAmount.setVisibility(View.GONE);
        tvEnterRemark.setVisibility(View.GONE);
        etRemark.setVisibility(View.VISIBLE);
        llReview.setVisibility(View.VISIBLE);
        etAmount.setEnabled(false);
    }

    @OnClick(R.id.btn_cancel_transfer)
    public void cancelTransfer() {
        getActivity().getSupportFragmentManager().popBackStack();
    }

    @OnClick(R.id.btn_add_beneficiary)
    public void addBeneficiary() {
        ((BaseActivity) getActivity()).replaceFragment(BeneficiaryAddOptionsFragment.newInstance(),
                true, R.id.container);
    }

    @OnClick(R.id.btn_try_again)
    public void onRetry() {
        if (Network.isConnected(getContext())) {
            sweetUIErrorHandler.hideSweetErrorLayoutUI(layoutMakeTransfer, layoutError);
            presenter.loadTransferTemplate();
        } else {
            Toaster.show(rootView, getString(R.string.internet_not_connected));
        }
    }

    /**
     * It is called whenever any error occurs while executing a request
     * @param msg Error message that tells the user about the problem.
     */
    @Override
    public void showError(String msg) {
        if (!Network.isConnected(getContext())) {
            sweetUIErrorHandler.showSweetNoInternetUI(layoutMakeTransfer, layoutError);
        } else {
            sweetUIErrorHandler.showSweetErrorUI(msg, layoutMakeTransfer, layoutError);
            Toaster.show(rootView, msg);
        }
    }

    @Override
    public void showProgress() {
        layoutMakeTransfer.setVisibility(View.GONE);
        showProgressBar();
    }

    @Override
    public void hideProgress() {
        hideProgressBar();
        layoutMakeTransfer.setVisibility(View.VISIBLE);
    }

    /**
     * Callback for {@code spPayFrom} and {@code spBeneficiary}
     * @param parent
     * @param view
     * @param position
     * @param id
     */
    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        switch (parent.getId()) {
            case R.id.sp_beneficiary:
                beneficiaryAccountOption = presenter.searchAccount(accountOptionsTemplate.
                        getFromAccountOptions(), beneficiaryAdapter.getItem(position));
                break;
            case R.id.sp_pay_from:
                fromAccountOption = accountOptionsTemplate.getFromAccountOptions().get(position);
                break;
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {

    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_transfer, menu);
        Utils.setToolbarIconColor(getActivity(), menu, R.color.white);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.menu_refresh_transfer) {
            FragmentTransaction transaction = getFragmentManager().beginTransaction();
            Fragment currFragment = getActivity().getSupportFragmentManager()
                    .findFragmentById(R.id.container);
            transaction.detach(currFragment);
            transaction.attach(currFragment);
            transaction.commit();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        hideProgress();
        presenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/ThirdPartyTransferFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.EditText
import android.widget.Spinner
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentThirdPartyTransferBinding
import org.mifos.mobile.models.beneficiary.Beneficiary
import org.mifos.mobile.models.beneficiary.BeneficiaryDetail
import org.mifos.mobile.models.payload.AccountDetail
import org.mifos.mobile.models.payload.TransferPayload
import org.mifos.mobile.models.templates.account.AccountOption
import org.mifos.mobile.models.templates.account.AccountOptionsTemplate
import org.mifos.mobile.presenters.ThirdPartyTransferPresenter
import org.mifos.mobile.ui.activities.base.BaseActivity
import org.mifos.mobile.ui.adapters.AccountsSpinnerAdapter
import org.mifos.mobile.ui.adapters.BeneficiarySpinnerAdapter
import org.mifos.mobile.ui.enums.TransferType
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.ui.views.ThirdPartyTransferView
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.DateHelper
import org.mifos.mobile.utils.Network
import org.mifos.mobile.utils.Toaster
import org.mifos.mobile.utils.Utils
import org.mifos.mobile.utils.getTodayFormatted
import javax.inject.Inject

/**
 * Created by dilpreet on 21/6/17.
 */
class ThirdPartyTransferFragment : BaseFragment(), ThirdPartyTransferView, OnItemSelectedListener {

    private var _binding: FragmentThirdPartyTransferBinding? = null
    private val binding get() = _binding!!

    @JvmField
    @Inject
    var presenter: ThirdPartyTransferPresenter? = null
    private val listBeneficiary: MutableList<BeneficiaryDetail?> = ArrayList()
    private val listPayFrom: MutableList<AccountDetail> = ArrayList()
    private var beneficiaries: List<Beneficiary?>? = null
    private var beneficiaryAdapter: BeneficiarySpinnerAdapter? = null
    private var payFromAdapter: AccountsSpinnerAdapter? = null
    private var fromAccountOption: AccountOption? = null
    private var beneficiaryAccountOption: AccountOption? = null
    private var accountOptionsTemplate: AccountOptionsTemplate? = null
    private var transferDate: String? = null
    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        (activity as BaseActivity?)?.activityComponent?.inject(this)
        _binding = FragmentThirdPartyTransferBinding.inflate(inflater, container, false)
        setToolbarTitle(getString(R.string.third_party_transfer))
        sweetUIErrorHandler = SweetUIErrorHandler(activity, binding.root)
        showUserInterface()
        presenter?.attachView(this)
        if (savedInstanceState == null) {
            presenter?.loadTransferTemplate()
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnReviewTransfer.setOnClickListener {
            reviewTransfer()
        }

        binding.btnPayFrom.setOnClickListener {
            payFromSelected()
        }

        binding.btnPayTo.setOnClickListener {
            payToSelected()
        }

        binding.btnAmount.setOnClickListener {
            amountSet()
        }

        binding.btnCancelTransfer.setOnClickListener {
            cancelTransfer()
        }

        binding.btnAddBeneficiary.setOnClickListener {
            addBeneficiary()
        }
        binding.layoutError.btnTryAgain.setOnClickListener {
            onRetry()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(Constants.TEMPLATE, accountOptionsTemplate)
        outState.putParcelableArrayList(
            Constants.BENEFICIARY,
            ArrayList<Parcelable?>(
                beneficiaries,
            ),
        )
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            showThirdPartyTransferTemplate(savedInstanceState.getParcelable<Parcelable>(Constants.TEMPLATE) as AccountOptionsTemplate)
            val tempBeneficiaries: List<Beneficiary?> = savedInstanceState.getParcelableArrayList(
                Constants.BENEFICIARY,
            ) ?: listOf()
            showBeneficiaryList(tempBeneficiaries)
        }
    }

    /**
     * Setting up basic components
     */
    override fun showUserInterface() {
        payFromAdapter = activity?.applicationContext?.let {
            AccountsSpinnerAdapter(it, listPayFrom)
        }
        payFromAdapter?.setDropDownViewResource(android.R.layout.select_dialog_singlechoice)
        binding.spPayFrom.adapter = payFromAdapter
        binding.spPayFrom.onItemSelectedListener = this
        beneficiaryAdapter = activity?.applicationContext?.let {
            BeneficiarySpinnerAdapter(
                it,
                R.layout.beneficiary_spinner_layout,
                listBeneficiary,
            )
        }
        beneficiaryAdapter?.setDropDownViewResource(android.R.layout.select_dialog_singlechoice)
        binding.spBeneficiary.adapter = beneficiaryAdapter
        binding.spBeneficiary.onItemSelectedListener = this
        transferDate = DateHelper.getSpecificFormat(
            DateHelper.FORMAT_dd_MMMM_yyyy,
            getTodayFormatted(),
        )
        binding.processOne.setCurrentActive()
    }

    /**
     * Checks validation of `etRemark` and then opens [TransferProcessFragment] for
     * initiating the transfer
     */
    fun reviewTransfer() {
        if (binding.etAmount.text.toString() == "") {
            Toaster.show(binding.root, getString(R.string.enter_amount))
            return
        }
        if (binding.etAmount.text.toString() == ".") {
            Toaster.show(binding.root, getString(R.string.invalid_amount))
            return
        }
        if (binding.etRemark.text.toString().trim { it <= ' ' } == "") {
            Toaster.show(binding.root, getString(R.string.remark_is_mandatory))
            return
        }
        if (binding.spBeneficiary.selectedItem.toString() == binding.spPayFrom.selectedItem.toString()) {
            Toaster.show(binding.root, getString(R.string.error_same_account_transfer))
            return
        }
        val transferPayload = TransferPayload()
        transferPayload.fromAccountId = fromAccountOption?.accountId
        transferPayload.fromClientId = fromAccountOption?.clientId
        transferPayload.fromAccountNumber = fromAccountOption?.accountNo
        transferPayload.fromAccountType = fromAccountOption?.accountType?.id
        transferPayload.fromOfficeId = fromAccountOption?.officeId
        transferPayload.toOfficeId = beneficiaryAccountOption?.officeId
        transferPayload.toAccountId = beneficiaryAccountOption?.accountId
        transferPayload.toClientId = beneficiaryAccountOption?.clientId
        transferPayload.toAccountNumber = beneficiaryAccountOption?.accountNo
        transferPayload.toAccountType = beneficiaryAccountOption?.accountType?.id
        transferPayload.transferDate = transferDate
        transferPayload.transferAmount = binding.etAmount.text.toString().toDouble()
        transferPayload.transferDescription = binding.etRemark.text.toString()
        transferPayload.fromAccountNumber = fromAccountOption?.accountNo
        transferPayload.toAccountNumber = beneficiaryAccountOption?.accountNo
        (activity as BaseActivity?)?.replaceFragment(
            TransferProcessFragment.newInstance(
                transferPayload,
                TransferType.TPT,
            ),
            true,
            R.id.container,
        )
    }

    /**
     * Shows a {@link Snackbar} with `message`
     *
     * @param msg String to be shown
     */
    override fun showToaster(msg: String?) {
        Toaster.show(binding.root, msg)
    }

    /**
     * Provides with `accountOptionsTemplate` fetched from server which is used to update
     * `listPayFrom`
     *
     * @param accountOptionsTemplate Template for account transfer
     */
    override fun showThirdPartyTransferTemplate(accountOptionsTemplate: AccountOptionsTemplate?) {
        this.accountOptionsTemplate = accountOptionsTemplate
        listPayFrom.clear()
        presenter?.getAccountNumbersFromAccountOptions(accountOptionsTemplate?.fromAccountOptions)
            ?.let { listPayFrom.addAll(it) }
        payFromAdapter?.notifyDataSetChanged()
    }

    /**
     * Provides with `beneficiaries` fetched from server which is used to update
     * `listBeneficiary`
     *
     * @param beneficiaries List of [Beneficiary] linked with user's account
     */
    override fun showBeneficiaryList(beneficiaries: List<Beneficiary?>?) {
        this.beneficiaries = beneficiaries
        listBeneficiary.clear()
        presenter?.getAccountNumbersFromBeneficiaries(beneficiaries)
            ?.let { listBeneficiary.addAll(it) }
        beneficiaryAdapter?.notifyDataSetChanged()
    }

    /**
     * Disables `spPayFrom` [Spinner] and sets `pvOne` to completed and make
     * `pvThree` pvTwo
     */
    fun payFromSelected() {
        binding.processOne.setCurrentCompleted()
        binding.processTwo.setCurrentActive()
        binding.btnPayFrom.visibility = View.GONE
        binding.tvSelectBeneficary.visibility = View.GONE
        if (listBeneficiary.isNotEmpty()) {
            binding.btnPayTo.visibility = View.VISIBLE
            binding.spBeneficiary.visibility = View.VISIBLE
            binding.spPayFrom.isEnabled = false
        } else {
            binding.tvAddBeneficiaryMsg.visibility = View.VISIBLE
            binding.btnAddBeneficiary.visibility = View.VISIBLE
        }
    }

    /**
     * Checks validation of `spBeneficiary` [Spinner].<br></br>
     * Disables `spBeneficiary` [Spinner] and sets `pvTwo` to completed and make
     * `pvThree` active
     */
    fun payToSelected() {
        if (binding.spBeneficiary.selectedItem.toString() == binding.spPayFrom.selectedItem.toString()) {
            showToaster(getString(R.string.error_same_account_transfer))
            return
        }
        binding.processTwo.setCurrentCompleted()
        binding.processThree.setCurrentActive()
        binding.btnPayTo.visibility = View.GONE
        binding.tvSelectAmount.visibility = View.GONE
        binding.etAmount.visibility = View.VISIBLE
        binding.btnAmount.visibility = View.VISIBLE
        binding.spBeneficiary.isEnabled = false
    }

    /**
     * Checks validation of `etAmount` [EditText].<br></br>
     * Disables `etAmount` and sets `pvThree` to completed and make
     * `pvFour` active
     */
    fun amountSet() {
        if (binding.etAmount.text.toString() == "") {
            showToaster(getString(R.string.enter_amount))
            return
        }
        if (binding.etAmount.text.toString() == ".") {
            showToaster(getString(R.string.invalid_amount))
            return
        }
        if (binding.etAmount.text.toString().toDouble() == 0.0) {
            showToaster(getString(R.string.amount_greater_than_zero))
            return
        }
        binding.processThree.setCurrentCompleted()
        binding.processFour.setCurrentActive()
        binding.btnAmount.visibility = View.GONE
        binding.tvEnterRemark.visibility = View.GONE
        binding.etRemark.visibility = View.VISIBLE
        binding.llReview.visibility = View.VISIBLE
        binding.etAmount.isEnabled = false
    }

    fun cancelTransfer() {
        activity?.supportFragmentManager?.popBackStack()
    }

    fun addBeneficiary() {
        (activity as BaseActivity?)?.replaceFragment(
            BeneficiaryAddOptionsFragment.newInstance(),
            true,
            R.id.container,
        )
    }

    fun onRetry() {
        if (Network.isConnected(context)) {
            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
                binding.llMakeTransfer,
                binding.layoutError.root,
            )
            presenter?.loadTransferTemplate()
        } else {
            Toaster.show(binding.root, getString(R.string.internet_not_connected))
        }
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param msg Error message that tells the user about the problem.
     */
    override fun showError(msg: String?) {
        if (!Network.isConnected(context)) {
            sweetUIErrorHandler?.showSweetNoInternetUI(
                binding.llMakeTransfer,
                binding.layoutError.root,
            )
        } else {
            sweetUIErrorHandler?.showSweetErrorUI(
                msg,
                binding.llMakeTransfer,
                binding.layoutError.root,
            )
        }
    }

    override fun showProgress() {
        binding.llMakeTransfer.visibility = View.GONE
        showProgressBar()
    }

    override fun hideProgress() {
        hideProgressBar()
        binding.llMakeTransfer.visibility = View.VISIBLE
    }

    /**
     * Callback for `spPayFrom` and `spBeneficiary`
     */
    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        when (parent?.id) {
            R.id.sp_beneficiary ->
                beneficiaryAccountOption =
                    presenter?.searchAccount(
                        accountOptionsTemplate?.toAccountOptions,
                        beneficiaryAdapter?.getItem(position),
                    )

            R.id.sp_pay_from ->
                fromAccountOption =
                    accountOptionsTemplate?.fromAccountOptions?.get(position)
        }
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}
    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_transfer, menu)
        Utils.setToolbarIconColor(activity, menu, R.color.white)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_refresh_transfer) {
            val transaction = fragmentManager?.beginTransaction()
            val currFragment = activity?.supportFragmentManager
                ?.findFragmentById(R.id.container)
            if (currFragment != null) {
                transaction?.detach(currFragment)
                transaction?.attach(currFragment)
            }
            transaction?.commit()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hideProgress()
        presenter?.detachView()
        _binding = null
    }

    companion object {
        fun newInstance(): ThirdPartyTransferFragment {
            return ThirdPartyTransferFragment()
        }
    }
}
